In [1]:
import pandas as pd

In [20]:
tfidf = pd.read_csv('./eval_data/evaluation_tfidf.csv', index_col=0)
bow = pd.read_csv('./eval_data/evaluation_bow.csv', index_col=0)
w2v = pd.read_csv('./eval_data/evaluation_w2v.csv', index_col=0)
ft = pd.read_csv('./eval_data/evaluation_ft.csv', index_col=0)

In [21]:
evaluation = pd.concat([tfidf,
                        bow,
                        w2v,
                        ft,
                        ])

In [22]:
evaluation

model variant vectorize_variant  train_acc  \
0                     StackingClassifier   clean             tfidf   0.995040   
1                 RandomForestClassifier   clean             tfidf   0.999897   
2              RandomForest-tuned-optuna   clean             tfidf   0.972305   
3         BalancedRandomForestClassifier   clean             tfidf   0.905549   
4   BalancedRandomForestClassifier-tuned   clean             tfidf   0.885398   
5                          XGBClassifier   clean             tfidf   0.959750   
6               XGBClassifier-tuned-grid   clean             tfidf   0.962333   
7               XGBClassifier-tuned-rand   clean             tfidf   0.951276   
8             XGBClassifier-tuned-optuna   clean             tfidf   0.978919   
9                     CatBoostClassifier   clean             tfidf   0.946058   
10                        LGBMClassifier   clean             tfidf   0.961713   
0                 RandomForestClassifier   clean               bow   0.999948   
1      RandomForestClassifier-tuned-rand   clean               bow   0.939857   
2              RandomForest-tuned-optuna   clean               bow   0.978196   
3         BalancedRandomForestClassifier   clean               bow   0.897541   
4   BalancedRandomForestClassifier-tuned   clean               bow   0.856670   
5                          XGBClassifier   clean               bow   0.956805   
6               XGBClassifier-tuned-grid   clean               bow   0.958200   
7               XGBClassifier-tuned-rand   clean               bow   0.941149   
8             XGBClassifier-tuned-optuna   clean               bow   0.990648   
9                     CatBoostClassifier   clean               bow   0.947039   
10                    StackingClassifier   clean               bow   0.998708   
0                 RandomForestClassifier   clean               w2v   0.999948   
1      RandomForestClassifier-tuned-rand   clean               w2v   0.943991   
2              RandomForest-tuned-optuna   clean               w2v   0.999897   
3               XGBClassifier-tuned-grid   clean               w2v   0.999948   
4               XGBClassifier-tuned-rand   clean               w2v   0.999173   
5                     CatBoostClassifier   clean               w2v   0.954170   
6                         LGBMClassifier   clean               w2v   0.995763   
7                     StackingClassifier   clean               w2v   0.969722   
8                          XGBClassifier   clean               w2v   0.999948   
9         BalancedRandomForestClassifier   clean               w2v   0.868916   
10  BalancedRandomForestClassifier-tuned   clean               w2v   0.826961   
0                 RandomForestClassifier   clean                ft   0.999948   
1      RandomForestClassifier-tuned-rand   clean                ft   0.943526   
2         BalancedRandomForestClassifier   clean                ft   0.860184   
3   BalancedRandomForestClassifier-tuned   clean                ft   0.817195   
4                          XGBClassifier   clean                ft   0.999948   
5               XGBClassifier-tuned-grid   clean                ft   0.999948   
6               XGBClassifier-tuned-rand   clean                ft   0.998192   
7             XGBClassifier-tuned-optuna   clean                ft   0.939651   
8                     CatBoostClassifier   clean                ft   0.952361   
9                         LGBMClassifier   clean                ft   0.995298   
10                    StackingClassifier   clean                ft   0.971065   
11             RandomForest-tuned-optuna   clean                ft   0.999948   

    train_prec  train_rec  train_f1  test_acc  test_prec  test_rec   test_f1  
0     0.988353   0.935433  0.961165  0.960699   0.799443  0.530499  0.637778  
1     1.000000   0.998425  0.999212  0.957806   0.832753  0.441774  0.577295  
2     0.990642   0.583465  0.734390  0.954430   0.865471  0.356747  0.5

In [23]:
evaluation.sort_values(by=["test_prec"], ascending=False)

model variant vectorize_variant  train_acc  \
9                     CatBoostClassifier   clean             tfidf   0.946058   
9                     CatBoostClassifier   clean               bow   0.947039   
2              RandomForest-tuned-optuna   clean             tfidf   0.972305   
2              RandomForest-tuned-optuna   clean               bow   0.978196   
7               XGBClassifier-tuned-rand   clean             tfidf   0.951276   
1                 RandomForestClassifier   clean             tfidf   0.999897   
5                          XGBClassifier   clean               bow   0.956805   
5                          XGBClassifier   clean             tfidf   0.959750   
0                     StackingClassifier   clean             tfidf   0.995040   
0                 RandomForestClassifier   clean                ft   0.999948   
0                 RandomForestClassifier   clean               w2v   0.999948   
1      RandomForestClassifier-tuned-rand   clean               w2v   0.943991   
6               XGBClassifier-tuned-grid   clean               bow   0.958200   
8             XGBClassifier-tuned-optuna   clean             tfidf   0.978919   
10                    StackingClassifier   clean               bow   0.998708   
11             RandomForest-tuned-optuna   clean                ft   0.999948   
6               XGBClassifier-tuned-grid   clean             tfidf   0.962333   
2              RandomForest-tuned-optuna   clean               w2v   0.999897   
1      RandomForestClassifier-tuned-rand   clean               bow   0.939857   
0                 RandomForestClassifier   clean               bow   0.999948   
10                        LGBMClassifier   clean             tfidf   0.961713   
6               XGBClassifier-tuned-rand   clean                ft   0.998192   
3               XGBClassifier-tuned-grid   clean               w2v   0.999948   
6                         LGBMClassifier   clean               w2v   0.995763   
5               XGBClassifier-tuned-grid   clean                ft   0.999948   
4               XGBClassifier-tuned-rand   clean               w2v   0.999173   
10                    StackingClassifier   clean                ft   0.971065   
7                     StackingClassifier   clean               w2v   0.969722   
1      RandomForestClassifier-tuned-rand   clean                ft   0.943526   
4                          XGBClassifier   clean                ft   0.999948   
8                          XGBClassifier   clean               w2v   0.999948   
8             XGBClassifier-tuned-optuna   clean               bow   0.990648   
9                         LGBMClassifier   clean                ft   0.995298   
5                     CatBoostClassifier   clean               w2v   0.954170   
7               XGBClassifier-tuned-rand   clean               bow   0.941149   
8                     CatBoostClassifier   clean                ft   0.952361   
7             XGBClassifier-tuned-optuna   clean                ft   0.939651   
3         BalancedRandomForestClassifier   clean             tfidf   0.905549   
4   BalancedRandomForestClassifier-tuned   clean             tfidf   0.885398   
3         BalancedRandomForestClassifier   clean               bow   0.897541   
4   BalancedRandomForestClassifier-tuned   clean               bow   0.856670   
2         BalancedRandomForestClassifier   clean                ft   0.860184   
9         BalancedRandomForestClassifier   clean               w2v   0.868916   
10  BalancedRandomForestClassifier-tuned   clean               w2v   0.826961   
3   BalancedRandomForestClassifier-tuned   clean                ft   0.817195   

    train_prec  train_rec  train_f1  test_acc  test_prec  test_rec   test_f1  
9     0.915441   0.196063  0.322957  0.945750   0.902655  0.188540  0.311927  
9     0.893891   0.218898  0.351676  0.946594   0.876923  0.210721  0.339791  
2     0.990642   0.583465  0.734390  0.954430   0.865471  0.356747  0.5

In [24]:
evaluation.sort_values(by=["test_rec"], ascending=False)

model variant vectorize_variant  train_acc  \
2         BalancedRandomForestClassifier   clean                ft   0.860184   
10  BalancedRandomForestClassifier-tuned   clean               w2v   0.826961   
3   BalancedRandomForestClassifier-tuned   clean                ft   0.817195   
9         BalancedRandomForestClassifier   clean               w2v   0.868916   
4   BalancedRandomForestClassifier-tuned   clean               bow   0.856670   
3         BalancedRandomForestClassifier   clean               bow   0.897541   
3         BalancedRandomForestClassifier   clean             tfidf   0.905549   
4   BalancedRandomForestClassifier-tuned   clean             tfidf   0.885398   
0                     StackingClassifier   clean             tfidf   0.995040   
8             XGBClassifier-tuned-optuna   clean               bow   0.990648   
10                    StackingClassifier   clean               bow   0.998708   
0                 RandomForestClassifier   clean               bow   0.999948   
1                 RandomForestClassifier   clean             tfidf   0.999897   
2              RandomForest-tuned-optuna   clean               bow   0.978196   
8             XGBClassifier-tuned-optuna   clean             tfidf   0.978919   
6               XGBClassifier-tuned-grid   clean               bow   0.958200   
2              RandomForest-tuned-optuna   clean             tfidf   0.972305   
8                          XGBClassifier   clean               w2v   0.999948   
4               XGBClassifier-tuned-rand   clean               w2v   0.999173   
6               XGBClassifier-tuned-grid   clean             tfidf   0.962333   
6                         LGBMClassifier   clean               w2v   0.995763   
5                          XGBClassifier   clean               bow   0.956805   
3               XGBClassifier-tuned-grid   clean               w2v   0.999948   
4                          XGBClassifier   clean                ft   0.999948   
6               XGBClassifier-tuned-rand   clean                ft   0.998192   
7                     StackingClassifier   clean               w2v   0.969722   
5               XGBClassifier-tuned-grid   clean                ft   0.999948   
9                         LGBMClassifier   clean                ft   0.995298   
5                          XGBClassifier   clean             tfidf   0.959750   
10                        LGBMClassifier   clean             tfidf   0.961713   
10                    StackingClassifier   clean                ft   0.971065   
5                     CatBoostClassifier   clean               w2v   0.954170   
8                     CatBoostClassifier   clean                ft   0.952361   
7               XGBClassifier-tuned-rand   clean             tfidf   0.951276   
9                     CatBoostClassifier   clean               bow   0.947039   
0                 RandomForestClassifier   clean                ft   0.999948   
2              RandomForest-tuned-optuna   clean               w2v   0.999897   
9                     CatBoostClassifier   clean             tfidf   0.946058   
0                 RandomForestClassifier   clean               w2v   0.999948   
11             RandomForest-tuned-optuna   clean                ft   0.999948   
7               XGBClassifier-tuned-rand   clean               bow   0.941149   
7             XGBClassifier-tuned-optuna   clean                ft   0.939651   
1      RandomForestClassifier-tuned-rand   clean                ft   0.943526   
1      RandomForestClassifier-tuned-rand   clean               w2v   0.943991   
1      RandomForestClassifier-tuned-rand   clean               bow   0.939857   

    train_prec  train_rec  train_f1  test_acc  test_prec  test_rec   test_f1  
2     0.319416   1.000000  0.484178  0.837734   0.259988  0.805915  0.393147  
10    0.253615   0.842520  0.389871  0.822905   0.239033  0.785582  0.366537  
3     0.240741   0.829134  0.373140  0.816998   0.232035  0.781885  0.3

In [25]:
evaluation.sort_values(by=["test_f1"], ascending=False)

model variant vectorize_variant  train_acc  \
0                     StackingClassifier   clean             tfidf   0.995040   
10                    StackingClassifier   clean               bow   0.998708   
8             XGBClassifier-tuned-optuna   clean               bow   0.990648   
1                 RandomForestClassifier   clean             tfidf   0.999897   
0                 RandomForestClassifier   clean               bow   0.999948   
2              RandomForest-tuned-optuna   clean               bow   0.978196   
8             XGBClassifier-tuned-optuna   clean             tfidf   0.978919   
2              RandomForest-tuned-optuna   clean             tfidf   0.972305   
6               XGBClassifier-tuned-grid   clean               bow   0.958200   
6               XGBClassifier-tuned-grid   clean             tfidf   0.962333   
5                          XGBClassifier   clean               bow   0.956805   
4               XGBClassifier-tuned-rand   clean               w2v   0.999173   
8                          XGBClassifier   clean               w2v   0.999948   
6                         LGBMClassifier   clean               w2v   0.995763   
3               XGBClassifier-tuned-grid   clean               w2v   0.999948   
6               XGBClassifier-tuned-rand   clean                ft   0.998192   
5                          XGBClassifier   clean             tfidf   0.959750   
4                          XGBClassifier   clean                ft   0.999948   
3         BalancedRandomForestClassifier   clean             tfidf   0.905549   
5               XGBClassifier-tuned-grid   clean                ft   0.999948   
4   BalancedRandomForestClassifier-tuned   clean             tfidf   0.885398   
7                     StackingClassifier   clean               w2v   0.969722   
3         BalancedRandomForestClassifier   clean               bow   0.897541   
10                        LGBMClassifier   clean             tfidf   0.961713   
9                         LGBMClassifier   clean                ft   0.995298   
10                    StackingClassifier   clean                ft   0.971065   
4   BalancedRandomForestClassifier-tuned   clean               bow   0.856670   
2         BalancedRandomForestClassifier   clean                ft   0.860184   
9         BalancedRandomForestClassifier   clean               w2v   0.868916   
5                     CatBoostClassifier   clean               w2v   0.954170   
7               XGBClassifier-tuned-rand   clean             tfidf   0.951276   
8                     CatBoostClassifier   clean                ft   0.952361   
10  BalancedRandomForestClassifier-tuned   clean               w2v   0.826961   
3   BalancedRandomForestClassifier-tuned   clean                ft   0.817195   
9                     CatBoostClassifier   clean               bow   0.947039   
0                 RandomForestClassifier   clean                ft   0.999948   
9                     CatBoostClassifier   clean             tfidf   0.946058   
2              RandomForest-tuned-optuna   clean               w2v   0.999897   
0                 RandomForestClassifier   clean               w2v   0.999948   
11             RandomForest-tuned-optuna   clean                ft   0.999948   
7               XGBClassifier-tuned-rand   clean               bow   0.941149   
7             XGBClassifier-tuned-optuna   clean                ft   0.939651   
1      RandomForestClassifier-tuned-rand   clean                ft   0.943526   
1      RandomForestClassifier-tuned-rand   clean               w2v   0.943991   
1      RandomForestClassifier-tuned-rand   clean               bow   0.939857   

    train_prec  train_rec  train_f1  test_acc  test_prec  test_rec   test_f1  
0     0.988353   0.935433  0.961165  0.960699   0.799443  0.530499  0.637778  
10    1.000000   0.980315  0.990060  0.958770   0.781870  0.510166  0.617450  
8     0.989218   0.866929  0.924045  0.953225   0.690773  0.512015  0.5

#### Auswertung

- Modelle, die eine gute Precision erzielen, haben einen schlechten Recall => die getroffenen Vorhersagen sind gut, allerdings werden nicht viele korrekt vorhergesagt (wenige true positives)
- Modelle, die einen guten Recall erzielen, haben einen schlechten Recall => es werden mehr korrekt vorhergesagt, allerdings bei allgemein schlechteren Vorhersagen (mehr false positives)
- deutlich erkennbar am f1 score => Modelle mit bester Precision/Recall mit schlechtem f1 score
- o.g. Punkte sowohl in Train als auch Test zu sehen
- bestes Modell: RandomForest ohne Tuning => ca. 81% Precision und 43% Recall bei Test, bei Train ~99%; cross validation (10) führt zu f1 von ~52%
- Performance RF deutet auf Overfitting hin, mittels RandomizedSearchCV jedoch kein Modell mit geringerer Train aber besserer Test Performance erreichbar (gilt auch für BalancedRandomForestClassifier)


Zielmetrik: F1-Score

XGBClassifier: GridSearch vs. RandomizedSearch
- Grid deutlich besser

StackingClassifier
- erzielt leicht bessere Ergebnisse als das beste Einzelmodell (SVC mit 61%, gestackt mit RandomForest 64%)